### PR TITLE
[DOCS] Add docs for designing resilient clusters

### PR DESCRIPTION
**Pull request by DaveCTurner**
This commit adds some guidance for designing clusters to be resilient to
failures, including illustrations and example architectures.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
This is WIP and the diagrams in particular are rough drafts. I expect a bunch of iterations.

**Comment by DaveCTurner** [comment]
Previews are available, particularly:

- http://elasticsearch_47233.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/high-availability-cluster-design.html
- http://elasticsearch_47233.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/_resilience_in_small_clusters.html
- http://elasticsearch_47233.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/_resilience_in_larger_clusters.html

**Comment by jrodewig** [comment]
@elasticmachine run elasticsearch-ci/docs

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+It is a good idea for both nodes to have all other roles, so that the other
+tasks in the cluster can be handled by either node. Nodes have all roles by
+default.
+
+You should avoid sending client requests to just one of your nodes. If you do,
+and this node fails, then any requests will not receive responses even if the
+remaining node is a healthy cluster on its own. Ideally you should balance your
+client requests across both nodes. A good way to do this is to specify the
+addresses of both nodes when configuring the client to connect to your cluster.
+Alternatively you can use a resilient load balancer to balance client requests
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:83]:
```suggestion
Alternatively, you can use a resilient load balancer to balance client requests
```

**Review by jrodewig — DISMISSED** [review]
Looks good overall. Thanks for putting this together!

Most of my feedback is minor, non-blocking rewording.

However, I think the information about cluster requirements is a little unclear and could be misread.
There are also a couple of broken links to Cloud in the summary sections.

**Review by jrodewig — APPROVED** [review]
LGTM. Made some wording changes and other corrections.

Also removed the diagrams as discussed.

I'll merge and backport once the CI passes.

**Comment by jrodewig** [comment]
### Backport commits
master 1f28bd0
7.x bc921ea 
7.8 a8c2076
7.7 1403cda

**Pull request merged**